// material-ui-native/lib/mixins/controllable.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reactNative = require('react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

exports['default'] = {

  propTypes: {
    onChange: _reactNative2['default'].PropTypes.func,
    value: _reactNative2['default'].PropTypes.oneOfType([_reactNative2['default'].PropTypes.string, _reactNative2['default'].PropTypes.array]),
    valueLink: _reactNative2['default'].PropTypes.shape({
      value: _reactNative2['default'].PropTypes.string.isRequired,
      requestChange: _reactNative2['default'].PropTypes.func.isRequired
    })
  },

  getDefaultProps: function getDefaultProps() {
    return {
      onChange: function onChange() {}
    };
  },

  getValueLink: function getValueLink(props) {
    return props.valueLink || {
      value: props.value,
      requestChange: props.onChange
    };
  }

};
module.exports = exports['default'];
