// material-ui-native/lib/lists/nested-list.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reactNative = require('../../../react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

var _list = require('./list');

var _list2 = _interopRequireDefault(_list);

var NestedList = _reactNative2['default'].createClass({
  displayName: 'NestedList',

  propTypes: {
    nestedLevel: _reactNative2['default'].PropTypes.number,
    open: _reactNative2['default'].PropTypes.bool,
    style: _reactNative2['default'].PropTypes.object
  },

  getDefaultProps: function getDefaultProps() {
    return {
      nestedLevel: 1,
      open: false
    };
  },

  render: function render() {
    var _props = this.props;
    var children = _props.children;
    var nestedLevel = _props.nestedLevel;
    var style = _props.style;

    return _reactNative2['default'].createElement(
      _list2['default'],
      { style: style },
      _reactNative2['default'].Children.map(children, function (child) {
        return _reactNative2['default'].isValidElement(child) ? _reactNative2['default'].cloneElement(child, {
          nestedLevel: nestedLevel + 1
        }) : child;
      })
    );
  }

});

exports['default'] = NestedList;
module.exports = exports['default'];
