// material-ui-native/lib/date-picker/calendar-month.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reactNative = require('../../../react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

var _utilsDateTime = require('../utils/date-time');

var _utilsDateTime2 = _interopRequireDefault(_utilsDateTime);

var _dayButton = require('./day-button');

var _dayButton2 = _interopRequireDefault(_dayButton);

var _clearfix = require('../clearfix');

var _clearfix2 = _interopRequireDefault(_clearfix);

var View = _reactNative2['default'].View;
var StyleSheet = _reactNative2['default'].StyleSheet;

var CalendarMonth = _reactNative2['default'].createClass({
  displayName: 'CalendarMonth',

  propTypes: {
    displayDate: _reactNative2['default'].PropTypes.object.isRequired,
    selectedDate: _reactNative2['default'].PropTypes.object.isRequired,
    autoOk: _reactNative2['default'].PropTypes.bool,
    maxDate: _reactNative2['default'].PropTypes.object,
    minDate: _reactNative2['default'].PropTypes.object,
    onDayTouchTap: _reactNative2['default'].PropTypes.func,
    shouldDisableDate: _reactNative2['default'].PropTypes.func
  },

  render: function render() {
    var styles = StyleSheet.create({
      lineHeight: '32px',
      textAlign: 'center',
      padding: '16px 14px 0 14px'
    });

    return _reactNative2['default'].createElement(
      View,
      { style: styles },
      this._getWeekElements()
    );
  },

  isSelectedDateDisabled: function isSelectedDateDisabled() {
    return this._selectedDateDisabled;
  },

  _getWeekElements: function _getWeekElements() {
    var _this = this;

    var weekArray = _utilsDateTime2['default'].getWeekArray(this.props.displayDate);

    return weekArray.map(function (week, i) {
      return _reactNative2['default'].createElement(
        _clearfix2['default'],
        { key: i },
        _this._getDayElements(week, i)
      );
    }, this);
  },

  _getDayElements: function _getDayElements(week, i) {
    var _this2 = this;

    return week.map(function (day, j) {
      var isSameDate = _utilsDateTime2['default'].isEqualDate(_this2.props.selectedDate, day);
      var disabled = _this2._shouldDisableDate(day);
      var selected = !disabled && isSameDate;

      if (isSameDate) {
        if (disabled) {
          _this2._selectedDateDisabled = true;
        } else {
          _this2._selectedDateDisabled = false;
        }
      }

      return _reactNative2['default'].createElement(_dayButton2['default'], {
        key: 'db' + i + j,
        date: day,
        onTouchTap: _this2._handleDayTouchTap,
        selected: selected,
        disabled: disabled });
    }, this);
  },

  _handleDayTouchTap: function _handleDayTouchTap(e, date) {
    if (this.props.onDayTouchTap) this.props.onDayTouchTap(e, date);
  },

  _shouldDisableDate: function _shouldDisableDate(day) {
    if (day === null) return false;
    var disabled = !_utilsDateTime2['default'].isBetweenDates(day, this.props.minDate, this.props.maxDate);
    if (!disabled && this.props.shouldDisableDate) disabled = this.props.shouldDisableDate(day);

    return disabled;
  }

});

exports['default'] = CalendarMonth;
module.exports = exports['default'];
