// material-ui-native/lib/date-picker/date-picker-inline.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _reactNative = require('../../../react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

var _paper = require('../paper');

var _paper2 = _interopRequireDefault(_paper);

var View = _reactNative2['default'].View;

var DatePickerInline = _reactNative2['default'].createClass({
  displayName: 'DatePickerInline',

  propTypes: {
    open: _reactNative2['default'].PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      open: false
    };
  },

  render: function render() {
    var _props = this.props;
    var actions = _props.actions;
    var children = _props.children;
    var open = _props.open;
    var style = _props.style;

    var other = _objectWithoutProperties(_props, ['actions', 'children', 'open', 'style']);

    if (!open) {
      return _reactNative2['default'].createElement(View, null);
    }

    var styles = {
      actions: {
        marginRight: 8,
        paddingBottom: 12,
        textAlign: 'right'
      },
      container: {
        //zIndex: 3,
        //width: '100%',
        position: 'relative'
      },
      //display: 'block',
      subContainer: {
        position: 'absolute',
        height: 'auto'
      }
    };
    return _reactNative2['default'].createElement(
      View,
      { style: styles.container },
      _reactNative2['default'].createElement(
        View,
        { style: styles.subContainer },
        _reactNative2['default'].createElement(
          _paper2['default'],
          other,
          children,
          _reactNative2['default'].createElement(
            View,
            { style: styles.actions },
            actions
          )
        )
      )
    );
  }

});

exports['default'] = DatePickerInline;
module.exports = exports['default'];
