// material-ui-native/lib/utils/prop-types.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _reactNative = require('react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

var horizontal = _reactNative2['default'].PropTypes.oneOf(['left', 'middle', 'right']);
var vertical = _reactNative2['default'].PropTypes.oneOf(['top', 'center', 'bottom']);

exports['default'] = {

  corners: _reactNative2['default'].PropTypes.oneOf(['bottom-left', 'bottom-right', 'top-left', 'top-right']),

  horizontal: horizontal,

  vertical: vertical,

  origin: _reactNative2['default'].PropTypes.shape({
    horizontal: horizontal,
    vertical: vertical
  }),

  cornersAndCenter: _reactNative2['default'].PropTypes.oneOf(['bottom-center', 'bottom-left', 'bottom-right', 'top-center', 'top-left', 'top-right']),

  stringOrNumber: _reactNative2['default'].PropTypes.oneOfType([_reactNative2['default'].PropTypes.string, _reactNative2['default'].PropTypes.number]),

  zDepth: _reactNative2['default'].PropTypes.oneOf([0, 1, 2, 3, 4, 5])

};
module.exports = exports['default'];
