// material-ui-native/lib/auto-complete.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _reactNative = require('../../react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

/*import ReactTransitionGroup from 'react-addons-transition-group';*/

var _mixinsStylePropable = require('./mixins/style-propable');

var _mixinsStylePropable2 = _interopRequireDefault(_mixinsStylePropable);

var _mixinsClickAwayable = require('./mixins/click-awayable');

var _mixinsClickAwayable2 = _interopRequireDefault(_mixinsClickAwayable);

var _utilsKeyCode = require('./utils/key-code');

var _utilsKeyCode2 = _interopRequireDefault(_utilsKeyCode);

var _textField = require('./text-field');

var _textField2 = _interopRequireDefault(_textField);

var _menusMenu = require('./menus/menu');

var _menusMenu2 = _interopRequireDefault(_menusMenu);

var _menusMenuItem = require('./menus/menu-item');

var _menusMenuItem2 = _interopRequireDefault(_menusMenuItem);

var _menusMenuDivider = require('./menus/menu-divider');

var _menusMenuDivider2 = _interopRequireDefault(_menusMenuDivider);

var View = _reactNative2['default'].View;
var StyleSheet = _reactNative2['default'].StyleSheet;

var AutoComplete = _reactNative2['default'].createClass({
  displayName: 'AutoComplete',

  mixins: [_mixinsStylePropable2['default'], _mixinsClickAwayable2['default']],

  contextTypes: {
    muiTheme: _reactNative2['default'].PropTypes.object
  },

  propTypes: {
    animated: _reactNative2['default'].PropTypes.bool,
    errorText: _reactNative2['default'].PropTypes.string,
    floatingLabelText: _reactNative2['default'].PropTypes.string,
    errorStyle: _reactNative2['default'].PropTypes.object,
    hintText: _reactNative2['default'].PropTypes.string,
    searchText: _reactNative2['default'].PropTypes.string,
    dataSource: _reactNative2['default'].PropTypes.array,
    updateWhenFocused: _reactNative2['default'].PropTypes.bool,
    showAllItems: _reactNative2['default'].PropTypes.bool,
    menuStyle: _reactNative2['default'].PropTypes.object,
    listStyle: _reactNative2['default'].PropTypes.object,
    menuProps: _reactNative2['default'].PropTypes.object,
    menuCloseDelay: _reactNative2['default'].PropTypes.number,
    onUpdateInput: _reactNative2['default'].PropTypes.func,
    onNewRequest: _reactNative2['default'].PropTypes.func,
    filter: _reactNative2['default'].PropTypes.func,
    disableFocusRipple: _reactNative2['default'].PropTypes.bool,
    fullWidth: _reactNative2['default'].PropTypes.bool,
    touchTapCloseDelay: _reactNative2['default'].PropTypes.number,
    open: _reactNative2['default'].PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      animated: true,
      fullWidth: false,
      open: false,
      showAllItems: false,
      searchText: '',
      menuCloseDelay: 100,
      disableFocusRipple: true,
      updateWhenFocused: false,
      onUpdateInput: function onUpdateInput() {},
      onNewRequest: function onNewRequest() {},
      filter: function filter(searchText, key) {
        return key.includes(searchText);
      }
    };
  },

  getInitialState: function getInitialState() {
    return {
      searchText: this.props.searchText,
      open: this.props.open
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (this.props.searchText !== nextProps.searchText) {
      this.setState({
        searchText: nextProps.searchText
      });
    }
  },
  componentWillMount: function componentWillMount() {
    this.focusOnInput = false;
    this.requestsList = [];
  },

  componentClickAway: function componentClickAway() {
    this.setState({ open: false });
    this.focusOnInput = false;
  },

  render: function render() {
    var _this = this;

    var _props = this.props;
    var animated = _props.animated;
    var style = _props.style;
    var errorStyle = _props.errorStyle;
    var floatingLabelText = _props.floatingLabelText;
    var hintText = _props.hintText;
    var fullWidth = _props.fullWidth;
    var menuStyle = _props.menuStyle;
    var menuProps = _props.menuProps;
    var listStyle = _props.listStyle;
    var showAllItems = _props.showAllItems;

    var other = _objectWithoutProperties(_props, ['animated', 'style', 'errorStyle', 'floatingLabelText', 'hintText', 'fullWidth', 'menuStyle', 'menuProps', 'listStyle', 'showAllItems']);

    var styles = StyleSheet.create({
      root: {
        //display: 'inline-block',
        position: 'relative'
      },
      //width: this.props.fullWidth ? '100%' : 256,
      input: {},
      error: {},
      menu: {
        top: this.props.floatingLabelText ? 64 : 40,
        left: 0
      },
      //width: '100%',
      list: {
        //display: 'block',
        //width: this.props.fullWidth ? '100%' : 256,
      }
    });

    var textFieldProps = {
      style: this.mergeAndPrefix(styles.input, style),
      floatingLabelText: floatingLabelText,
      hintText: !hintText && !floatingLabelText ? '' : hintText,
      fullWidth: true,
      multiLine: false,
      errorStyle: this.mergeAndPrefix(styles.error, errorStyle)
    };

    var mergedRootStyles = this.mergeAndPrefix(styles.root, style);
    var mergedMenuStyles = this.mergeStyles(styles.menu, menuStyle);

    var displayFilter = showAllItems ? function () {
      return true;
    } : this.props.filter;
    var requestsList = [];

    this.props.dataSource.map(function (item) {
      switch (typeof item) {
        case 'string':
          if (displayFilter(_this.state.searchText, item, item)) {
            requestsList.push(item);
          }
          break;
        case 'object':
          if (typeof item.text === 'string') {
            if (displayFilter(_this.state.searchText, item.text, item.value)) {
              requestsList.push(item);
            } else if (item.display) {
              requestsList.push(item);
            }
          }
          break;
      }
    });

    this.requestsList = requestsList;

    var menu = this.state.open && (this.state.searchText !== '' || showAllItems) && requestsList.length > 0 ? _reactNative2['default'].createElement(
      _menusMenu2['default'],
      _extends({}, menuProps, {
        ref: 'menu',
        key: 'dropDownMenu',
        animated: animated,
        autoWidth: false,
        initiallyKeyboardFocused: false,
        onEscKeyDown: function () {
          return _this.setState({ open: false });
        },
        onItemTouchTap: this._handleItemTouchTap,
        listStyle: this.mergeAndPrefix(styles.list, listStyle),
        openDirection: 'bottom-left',
        style: mergedMenuStyles }),
      requestsList.map(function (request, index) {
        switch (typeof request) {
          case 'string':
            return _reactNative2['default'].createElement(_menusMenuItem2['default'], {
              disableFocusRipple: _this.props.disableFocusRipple,
              innerDivStyle: { overflow: 'hidden' },
              key: index,
              value: request,
              primaryText: request
            });
          case 'object':
            if (typeof request.text === 'string') {
              return _reactNative2['default'].cloneElement(request.value, {
                key: request.text,
                disableFocusRipple: _this.props.disableFocusRipple
              });
            }
            return _reactNative2['default'].cloneElement(request, {
              key: index,
              disableFocusRipple: _this.props.disableFocusRipple
            });
          default:
            return null;
        }
      })
    ) : null;

    return _reactNative2['default'].createElement(
      View,
      { style: mergedRootStyles,
        onKeyDown: this._handleKeyDown },
      _reactNative2['default'].createElement(
        View,
        {
          style: {
            //width:'100%',
          } },
        _reactNative2['default'].createElement(_textField2['default'], _extends({}, other, {
          ref: 'searchTextField',
          value: this.state.searchText,
          onEnterKeyDown: function () {
            setTimeout(function () {
              _this.setState({ open: false });
            }, _this.props.touchTapCloseDelay);
            _this.props.onNewRequest(_this.state.searchText);
          },
          onChange: function (e) {
            var searchText = e.target.value;
            _this._updateRequests(searchText);
          },
          onBlur: function () {
            if (_this.focusOnInput && _this.state.open) _this.refs.searchTextField.focus();
          },
          onFocus: function () {
            if (!_this.state.open && (showAllItems || _this.props.updateWhenFocused || _this.state.searchText !== '')) {
              _this._updateRequests(_this.state.searchText);
            }
            _this.focusOnInput = true;
          }

        }, textFieldProps))
      )
    );
  },
  // TODO: <ReactTransitionGroup>{menu}</ReactTransitionGroup>

  setValue: function setValue(textValue) {
    this.setState({
      searchText: textValue
    });
  },

  getValue: function getValue() {
    return this.state.searchText;
  },

  _updateRequests: function _updateRequests(searchText) {

    this.setState({
      searchText: searchText,
      open: true
    });

    this.focusOnInput = true;

    this.props.onUpdateInput(searchText, this.props.dataSource);
  },

  _handleItemTouchTap: function _handleItemTouchTap(e, child) {
    var _this2 = this;

    setTimeout(function () {
      _this2.setState({ open: false });
    }, this.props.touchTapCloseDelay);

    var dataSource = this.props.dataSource;

    var chosenRequest = undefined,
        index = undefined,
        searchText = undefined;
    if (typeof dataSource[0] === 'string') {
      chosenRequest = this.requestsList[parseInt(child.key, 10)];
      index = dataSource.indexOf(chosenRequest);
      searchText = dataSource[index];
    } else {
      chosenRequest = child.key;
      index = dataSource.indexOf(dataSource.filter(function (item) {
        return chosenRequest === item.text;
      })[0]);
      searchText = chosenRequest;
    }

    this.setState({ searchText: searchText });

    this.props.onNewRequest(chosenRequest, index, dataSource);
  },

  _handleKeyDown: function _handleKeyDown(e) {
    switch (e.keyCode) {
      case _utilsKeyCode2['default'].ESC:
        this.setState({ open: false });
        break;
      case _utilsKeyCode2['default'].DOWN:
        if (this.focusOnInput && this.state.open) {
          e.preventDefault();
          this.focusOnInput = false;
          this.setState({ open: true });
        }
        break;
      default:
        break;
    }
  }

});

AutoComplete.Item = _menusMenuItem2['default'];
AutoComplete.Divider = _menusMenuDivider2['default'];

exports['default'] = AutoComplete;
module.exports = exports['default'];
