// material-ui-native/lib/ripples/circle-ripple.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var _reactNative = require('../../../react-native');

var _reactNative2 = _interopRequireDefault(_reactNative);

/*import ReactDOM from 'react-dom';*/
/*import PureRenderMixin from 'react-addons-pure-render-mixin';*/

var _mixinsStylePropable = require('../mixins/style-propable');

var _mixinsStylePropable2 = _interopRequireDefault(_mixinsStylePropable);

/*import AutoPrefix from '../styles/auto-prefix';*/

var _stylesTransitions = require('../styles/transitions');

var _stylesTransitions2 = _interopRequireDefault(_stylesTransitions);

var _stylesColors = require('../styles/colors');

var _stylesColors2 = _interopRequireDefault(_stylesColors);

var View = _reactNative2['default'].View;

var CircleRipple = _reactNative2['default'].createClass({
  displayName: 'CircleRipple',

  mixins: [/*PureRenderMixin, */_mixinsStylePropable2['default']],

  propTypes: {
    color: _reactNative2['default'].PropTypes.string,
    opacity: _reactNative2['default'].PropTypes.number,
    style: _reactNative2['default'].PropTypes.object
  },

  getDefaultProps: function getDefaultProps() {
    return {
      color: _stylesColors2['default'].darkBlack,
      opacity: 0.16
    };
  },

  componentWillAppear: function componentWillAppear(callback) {
    this._initializeAnimation(callback);
  },

  componentWillEnter: function componentWillEnter(callback) {
    this._initializeAnimation(callback);
  },

  componentDidAppear: function componentDidAppear() {
    this._animate();
  },

  componentDidEnter: function componentDidEnter() {
    this._animate();
  },

  componentWillLeave: function componentWillLeave(callback) {
    var _this = this;

    /*    let style = ReactDOM.findDOMNode(this).style;*/
    style.opacity = 0;
    setTimeout(function () {
      if (_this.isMounted()) callback();
    }, 2000);
  },

  render: function render() {
    var _props = this.props;
    var color = _props.color;
    var opacity = _props.opacity;
    var style = _props.style;

    var other = _objectWithoutProperties(_props, ['color', 'opacity', 'style']);

    var mergedStyles = this.mergeAndPrefix({
      position: 'absolute',
      top: 0,
      left: 0,
      //height: '100%',
      //width: '100%',
      //TODO: borderRadius: '50%',
      backgroundColor: color
    }, style);

    return _reactNative2['default'].createElement(View, _extends({}, other, { style: mergedStyles }));
  },

  _animate: function _animate() {
    /*    let style = ReactDOM.findDOMNode(this).style;*/
    var transitionValue = _stylesTransitions2['default'].easeOut('2s', 'opacity') + ',' + _stylesTransitions2['default'].easeOut('1s', 'transform');
    /*    AutoPrefix.set(style, 'transition', transitionValue);*/
    /*    AutoPrefix.set(style, 'transform', 'scale(1)');*/
  },

  _initializeAnimation: function _initializeAnimation(callback) {
    var _this2 = this;

    /*    let style = ReactDOM.findDOMNode(this).style;*/
    style.opacity = this.props.opacity;
    /*    AutoPrefix.set(style, 'transform', 'scale(0)');*/
    setTimeout(function () {
      if (_this2.isMounted()) callback();
    }, 0);
  }

});

exports['default'] = CircleRipple;
module.exports = exports['default'];
